llvm package fails to install under FreeBSD

Hi, folks. Here's what I get after `opam install llvm`:
```
The following actions will be performed:
  ∗  install llvm 3.9

=-=- Gathering sources =-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
[llvm] Archive in cache

=-=- Processing actions -=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
[ERROR] The compilation of llvm failed at "bash -ex ./build.sh 3.9
        ~/.opam/4.03.0+flambda/lib".
Processing  1/1: [llvm: sh]
#=== ERROR while installing llvm.3.9 ==========================================#
# opam-version 1.2.2
# os           freebsd
# command      bash -ex ./build.sh 3.9 ~/.opam/4.03.0+flambda/lib
# path         ~/.opam/4.03.0+flambda/build/llvm.3.9
# compiler     4.03.0+flambda
# exit-code    1
# env-file     ~/.opam/4.03.0+flambda/build/llvm.3.9/llvm-13291-6d07ae.env
# stdout-file  ~/.opam/4.03.0+flambda/build/llvm.3.9/llvm-13291-6d07ae.out
# stderr-file  ~/.opam/4.03.0+flambda/build/llvm.3.9/llvm-13291-6d07ae.err
### stdout ###
# 77 33 p--------- 0 magv magv 0 0 "/usr/local/Cellar/llvm/3.9*/bin/llvm-config" "/usr/local/Cellar/llvm/3.9*/bin/llvm-config" "/usr/local/Cellar/llvm/3.9*/bin/llvm-config" "/usr/local/Cellar/llvm/3.9*/bin/llvm-config" 4096 0 0 (stdin)
# Error: LLVM is not installed.
### stderr ###
# [...]
# ++ llvm-config-mp-3.9 --version
# ./build.sh: line 14: llvm-config-mp-3.9: command not found
# + continue
# + for llvm_config in 'llvm-config-$version' 'llvm-config-mp-$version' '$brew_llvm_config' llvm-config
# + case `$llvm_config --version` in
# ++ llvm-config --version
# ./build.sh: line 14: llvm-config: command not found
# + continue
# + echo 'Error: LLVM is not installed.'
# + exit 1
```

The immediate problem being that [this loop from build.sh](//github.com/ocaml/opam-repository/blob/master/packages/llvm/llvm.3.9/files/build.sh#L13):

```shell
for llvm_config in llvm-config-$version llvm-config-mp-$version $brew_llvm_config llvm-config; do
```

... does not include the actual name of `llvm-config` binary on FreeBSD, which is `llvm-config39`.

* * *

Unfortunately my attempts to fix this line on my machine and re-run the build are unsuccessful: editing `~/.opam/repo/default/packages/llvm/llvm.3.9/files/build.sh` and `~/.opam/4.03.0+flambda/build/llvm.3.9/build.sh` doesn't make a difference because `opam install` seems to ignore the first and override the second.

If anyone could advise how to actually edit a package locally to attempt to fix it, I'd appreciate that too. Creating a local repo with `opam repo add ...`, copying `~/.opam/repo/default/packages/llvm/llvm.3.9` there under a new name and trying to install that results in this error:
```
=-=- Processing actions -=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
[ERROR] Patch file "~/.opam/4.03.0+flambda/build/llvm-x.3.9/D24354.diff" not found.
[ERROR] Patch file "~/.opam/4.03.0+flambda/build/llvm-x.3.9/cmake.patch" not found.
```

There's an additional problem: even if the patch above may fix the building of llvm package, using it is still impossible and results in this sort of error messages:
```
+ ocamlfind ocamlopt -linkpkg -package llvm,llvm.analysis,llvm.bitwriter,... file.cmx ... -o file.native
/usr/local/bin/ld: cannot find -lLLVMBitWriter
/usr/local/bin/ld: cannot find -lLLVMAnalysis
/usr/local/bin/ld: cannot find -lLLVMProfileData
/usr/local/bin/ld: cannot find -lLLVMCore
/usr/local/bin/ld: cannot find -lLLVMSupport
/usr/local/bin/ld: cannot find -lLLVMAnalysis
/usr/local/bin/ld: cannot find -lLLVMProfileData
/usr/local/bin/ld: cannot find -lLLVMCore
/usr/local/bin/ld: cannot find -lLLVMSupport
/usr/local/bin/ld: cannot find -lLLVMCore
/usr/local/bin/ld: cannot find -lLLVMSupport
collect2: error: ld returned 1 exit status
File "caml_startup", line 1:
Error: Error during linking
Command exited with code 2.
```

These libraries are all present in `/usr/local/llvm39/lib`, but it seems that this path is not added to any of the CMA files? Is there a way for the package to add this library path automatically, or do users need to specify that manually?

* * *

Finally, even if you add that patch manually (by e.g. adding ` -lflags -cclib,-L/usr/local/llvm39/lib` option to ocamlbuild), you still can't use the llvm package, because building with it results in a wall of error messages like ``undefined reference to `std::__1::system_category()'``; adding `-lflags -cclib,-lc++` to ocamlbuild arguments fixes this. Again, is there a way for the package to do this automatically?

Yes, I'm using ocamlbuild. Adding `-lflags -linkall` to ocamlbuild arguments still results in the same error:
```
$ ocamlbuild -no-links -use-ocamlfind -r \
    -lflags -linkall \
    -tag-line "<*>: package(llvm,llvm.analysis,llvm.bitwriter)"
+ ocamlfind ocamlopt -linkall -linkpkg -package llvm,...
/usr/local/bin/ld: cannot find -lLLVMBitWriter
/usr/local/bin/ld: cannot find -lLLVMAnalysis
...
```

@jpdeplaix, `llvm-config39 --shared-mode` is static. If I force `link.patch` application the error changes to `/usr/local/bin/ld: cannot find -lLLVM-3.9`.

@whitequark, here's the list: [llvm39.files.txt](https://github.com/ocaml/opam-repository/files/714532/llvm39.files.txt); you can also more or less see it in `pkg-plist` file [over here](https://svnweb.freebsd.org/ports/head/devel/llvm39/).

--- a/packages/llvm/llvm.3.4/files/compile.sh
+++ b/packages/llvm/llvm.3.4/files/compile.sh
@@ -40,7 +40,8 @@ then
     brew_llvm_config=
 fi
 
-for config in llvm-config-$version llvm-config-mp-$version $brew_llvm_config llvm-config; do
+version_sans_dot=$(echo $version | tr -d .)
+for config in llvm-config-$version llvm-config$version_sans_dot llvm-config-mp-$version $brew_llvm_config llvm-config; do
     case `"$config" --version` in
         $version|$version.*)
             configure
@@ -53,5 +54,5 @@ for config in llvm-config-$version llvm-config-mp-$version $brew_llvm_config llv
     esac
 done
 
-echo "Error: LLVM is not installed."
+echo "Error: LLVM $version is not installed."
 exit 1

--- a/packages/llvm/llvm.3.5/files/compile.sh
+++ b/packages/llvm/llvm.3.5/files/compile.sh
@@ -40,7 +40,8 @@ then
     brew_llvm_config=
 fi
 
-for config in llvm-config-$version llvm-config-mp-$version $brew_llvm_config llvm-config; do
+version_sans_dot=$(echo $version | tr -d .)
+for config in llvm-config-$version llvm-config$version_sans_dot llvm-config-mp-$version $brew_llvm_config llvm-config; do
     case `"$config" --version` in
         $version|$version.*)
             configure
@@ -53,5 +54,5 @@ for config in llvm-config-$version llvm-config-mp-$version $brew_llvm_config llv
     esac
 done
 
-echo "Error: LLVM is not installed."
+echo "Error: LLVM $version is not installed."
 exit 1

--- a/packages/llvm/llvm.3.6/files/install.sh
+++ b/packages/llvm/llvm.3.6/files/install.sh
@@ -27,7 +27,8 @@ fi
 
 shopt -s nullglob
 
-for config in llvm-config-$version llvm-config-mp-$version $brew_llvm_config llvm-config; do
+version_sans_dot=$(echo $version | tr -d .)
+for config in llvm-config-$version llvm-config$version_sans_dot llvm-config-mp-$version $brew_llvm_config llvm-config; do
     case `$config --version` in
         $version|$version.*)
             configure
@@ -44,5 +45,5 @@ for config in llvm-config-$version llvm-config-mp-$version $brew_llvm_config llv
     esac
 done
 
-echo "Error: LLVM is not installed."
+echo "Error: LLVM $version is not installed."
 exit 1

--- a/packages/llvm/llvm.3.7/files/install.sh
+++ b/packages/llvm/llvm.3.7/files/install.sh
@@ -27,7 +27,8 @@ fi
 
 shopt -s nullglob
 
-for config in llvm-config-$version llvm-config-mp-$version $brew_llvm_config llvm-config; do
+version_sans_dot=$(echo $version | tr -d .)
+for config in llvm-config-$version llvm-config$version_sans_dot llvm-config-mp-$version $brew_llvm_config llvm-config; do
     case `$config --version` in
         $version|$version.*)
             mkdir build
@@ -46,5 +47,5 @@ for config in llvm-config-$version llvm-config-mp-$version $brew_llvm_config llv
     esac
 done
 
-echo "Error: LLVM is not installed."
+echo "Error: LLVM $version is not installed."
 exit 1

--- a/packages/llvm/llvm.3.8/files/install.sh
+++ b/packages/llvm/llvm.3.8/files/install.sh
@@ -27,7 +27,8 @@ fi
 
 shopt -s nullglob
 
-for config in llvm-config-$version llvm-config-mp-$version $brew_llvm_config llvm-config; do
+version_sans_dot=$(echo $version | tr -d .)
+for config in llvm-config-$version llvm-config$version_sans_dot llvm-config-mp-$version $brew_llvm_config llvm-config; do
     case `$config --version` in
         $version|$version.*)
             mkdir build
@@ -46,5 +47,5 @@ for config in llvm-config-$version llvm-config-mp-$version $brew_llvm_config llv
     esac
 done
 
-echo "Error: LLVM is not installed."
+echo "Error: LLVM $version is not installed."
 exit 1

--- a/packages/llvm/llvm.3.9/files/build.sh
+++ b/packages/llvm/llvm.3.9/files/build.sh
@@ -10,7 +10,8 @@ fi
 
 shopt -s nullglob
 
-for llvm_config in llvm-config-$version llvm-config-mp-$version $brew_llvm_config llvm-config; do
+version_sans_dot=$(echo $version | tr -d .)
+for llvm_config in llvm-config-$version llvm-config$version_sans_dot llvm-config-mp-$version $brew_llvm_config llvm-config; do
     case `$llvm_config --version` in
         $version|$version.*)
             case `$llvm_config --shared-mode` in
@@ -31,5 +32,5 @@ for llvm_config in llvm-config-$version llvm-config-mp-$version $brew_llvm_confi
     esac
 done
 
-echo "Error: LLVM is not installed."
+echo "Error: LLVM $version is not installed."
 exit 1

